File: 140. Word Break II.rb
#* 140. Word Break II
s = "catsanddog", word_dict = ["cat","cats","and","sand","dog"]
# @param {String} s
# @param {String[]} word_dict
# @return {String[]}
#*
def word_break(s, word_dict)
    f = lambda do |i|
    result = word_dict.include?(s[i..-1]) ? [s[i..-1]] : []
    (i+1...s.length).each do |j|
        if word_dict.include?(s[i...j])
        f.call(j).each do |t|
            result << s[i...j] + ' ' + t
        end
        end
    end
    result
    end

    f.call(0)
end
p word_break(s, word_dict)
  

File: 140. Word Break II.js
// 140. Word Break II
s = "catsanddog", wordDict = ["cat","cats","and","sand","dog"]
/**
 * @param {string} s
 * @param {string[]} wordDict
 * @return {string[]}
 */
var wordBreak = function(s, wordDict) {
    const f = (i) => {
        let result = wordDict.includes(s.slice(i)) ? [s.slice(i)] : [];
        for (let j = i + 1; j < s.length; j++) {
            if (wordDict.includes(s.slice(i, j))) {
                for (let t of f(j)) {
                    result.push(s.slice(i, j) + ' ' + t);
                }
            }
        }
        return result;
    };

    return f(0);
};
console.log(wordBreak(s,wordDict))

